code/datums/components/server.dm

```diff
@@ -0,0 +1,69 @@
+// Server component
+// will generate a heat based on the power usage of the machine
+// use only with /obj/machinery
+
+
+
+/datum/component/server
+
+	var/efficiency = 1 // 0 to 1 range
+
+	var/temperature = T20C // current temperature
+	var/warning_temp = T0C + 50 // 50C
+	var/overheated_temp = T0C + 100 // 100C - temperature at which the server stops working
+	var/heat_capacity = 2500 // Used for auxmos heat transfer
+
+	var/heat_generation = 0 // heat generated by the machine
+
+	// if it sparks when overheated
+	var/sparks = TRUE
+	COOLDOWN_DECLARE(spark_cooldown)
+	var/sparks_cooldown_time = 10 SECONDS
+
+/datum/component/server/Initialize()
+	if(!ismachinery(parent)) // currently only compatible with machinery
+		return COMPONENT_INCOMPATIBLE
+	RegisterSignal(parent, COMSIG_MACHINERY_POWER_USED, PROC_REF(ParentPowerUsed))
+	START_PROCESSING(SSservers, src)
+
+/datum/component/server/Destroy(force, silent)
+	. = ..()
+	UnregisterSignal(parent, COMSIG_MACHINERY_POWER_USED)
+	STOP_PROCESSING(SSservers, src)
+
+/datum/component/server/proc/ParentPowerUsed(source, amount, chan)
+	heat_generation += amount * 1000
+
+// server is overheated and doesn't work
+/datum/component/server/proc/overheated(is_overheated)
+	if(is_overheated && sparks && COOLDOWN_FINISHED(src, spark_cooldown))
+		do_sparks(5, FALSE, parent)
+		COOLDOWN_START(src, spark_cooldown, sparks_cooldown_time)
+
+	SEND_SIGNAL(parent, COMSIG_MACHINERY_OVERHEAT_CHANGE, is_overheated)
+
+/datum/component/server/process(delta_time)
+	var/obj/machinery/parent_machine = parent
+	calculate_temperature()
```

Comment:
I think it should do early return when it does `return FALSE`(line 66) because it means there's no turf, and processing this might not be a good idea.

Reply:
done